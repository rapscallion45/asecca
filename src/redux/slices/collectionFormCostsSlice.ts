import {
  ActionReducerMapBuilder,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from '@reduxjs/toolkit';
import {
  ICollectionFormCostsData,
  ICollectionFormCostsDataPayload,
} from '@/lib/api/api-types';
import collectionFormService from '../../services/forms/collectionFormService';
import {
  ICollectionFormCostsState,
  IFetchCollectionFormCostsByCollectionIdArgs,
  ISaveCollectionFormCostsByCollectionIdArgs,
  ICollectionFormCostsEditCostsPayload,
} from '../types';
import { addNotification } from './notificationsSlice';

/**
 * State slice definition for Collection Form Costs
 *
 * All state updates to the Collection Form Costs are managed here.
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 */

/**
 * Async thunk for GET /api/collection/costs/api/costs API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof CollectionFormCostsReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const fetchByCollectionId = createAsyncThunk(
  'collectionFormCosts/fetchByCollectionId',
  async (args: IFetchCollectionFormCostsByCollectionIdArgs, thunkAPI) => {
    /* await the result from the GET request */
    const res = await collectionFormService.getCosts(args.collectionId);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to load Collection Form Costs from server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Async thunk for POST /api/collection/costs/api/costs API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof CollectionFormCostsReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const saveByCollectionId = createAsyncThunk(
  'collectionFormCosts/saveByCollectionId',
  async (args: ISaveCollectionFormCostsByCollectionIdArgs, thunkAPI) => {
    const res = await collectionFormService.setCosts(args.data);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to save Collection Form Costs to server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    } else {
      thunkAPI.dispatch(
        addNotification({
          message: 'Successfully saved Collection Form Costs to server',
          variant: 'success',
        })
      );
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Initialises Collection Form Costs state to empty array
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof CollectionFormCostsReduxSlice
 *
 * @constant
 * @type {ICollectionFormCostsState}
 */
const initialCollectionFormCostsState: ICollectionFormCostsState = {
  loading: false,
  data: { rows: [] },
  dataShadow: { rows: [] },
  saving: false,
  edited: false,
};

/**
 * Create the redux slice for interacting with the Collection Form Costs state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof CollectionFormCostsReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {Slice<Object>}
 */
const collectionFormCostsSlice = createSlice({
  name: 'collectionFormCosts',
  initialState: initialCollectionFormCostsState,
  reducers: {
    /* reducer used for user input changes to the Costs data */
    editCosts: (
      state,
      action: PayloadAction<ICollectionFormCostsEditCostsPayload>
    ) => {
      /* find and update passed cost */
      state.data.rows = state.data.rows.map(
        (cost: ICollectionFormCostsData, index: number) => {
          /* perform update for passed table row number */
          if (index === action.payload.rowIdx) {
            return {
              ...cost,
              /* update the value of the passed column */
              [action.payload.colKey as keyof ICollectionFormCostsData]:
                action.payload.value,
            };
          }
          return cost;
        }
      );
      state.edited = true;
    },
    /* reducer used for when user clears edits to Costs data */
    resetCosts: (state) => {
      /* reset the data by simply copying the shadow to working copy */
      state.data = state.dataShadow;
      state.edited = false;
    },
  },
  extraReducers: (
    builder: ActionReducerMapBuilder<ICollectionFormCostsState>
  ) => {
    /*
     * `extraReducers` used to handle actions that were generated
     * outside of the Collection Form costs slice, such as the async thunks
     * for the API requests, defined at the top of this file.
     */
    builder
      /* Fetch Collection Form Costs extra reducers */
      .addCase(
        fetchByCollectionId.pending,
        (state: ICollectionFormCostsState) => {
          state.loading = true;
          state.data = { rows: [] };
          state.dataShadow = { rows: [] };
          state.error = undefined;
          state.edited = false;
        }
      )
      .addCase(
        fetchByCollectionId.fulfilled,
        (
          state: ICollectionFormCostsState,
          action: PayloadAction<ICollectionFormCostsDataPayload>
        ) => {
          state.loading = false;
          state.data = action.payload;
          state.dataShadow = action.payload;
          state.error = undefined;
        }
      )
      .addCase(
        fetchByCollectionId.rejected,
        (state: ICollectionFormCostsState) => {
          state.loading = false;
          state.data = { rows: [] };
          state.dataShadow = { rows: [] };
          state.error =
            'Failed to load Collection Form Costs data from server.';
        }
      )
      /* Save Collection Form Costs extra reducers */
      .addCase(
        saveByCollectionId.pending,
        (state: ICollectionFormCostsState) => {
          state.saving = true;
        }
      )
      .addCase(
        saveByCollectionId.fulfilled,
        (state: ICollectionFormCostsState) => {
          state.saving = false;
          state.edited = false;
        }
      )
      .addCase(
        saveByCollectionId.rejected,
        (state: ICollectionFormCostsState) => {
          state.saving = false;
        }
      );
  },
});

/* Collection Form Costs actions for editing and resetting costs data */
export const { editCosts, resetCosts } = collectionFormCostsSlice.actions;

export default collectionFormCostsSlice.reducer;
