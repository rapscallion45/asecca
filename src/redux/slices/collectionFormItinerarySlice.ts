import {
  ActionReducerMapBuilder,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from '@reduxjs/toolkit';
import {
  ICollectionFormItineraryData,
  ICollectionFormItineraryDataPayload,
  ICollectionFormItineraryAssetCategoryDataPayload,
} from '@/lib/api/api-types';
import collectionFormService from '../../services/forms/collectionFormService';
import {
  ICollectionFormItineraryState,
  IFetchCollectionFormItineraryByCollectionIdArgs,
  ISaveCollectionFormItineraryByCollectionIdArgs,
  ICollectionFormItineraryEditItineraryPayload,
  ICollectionFormItineraryDeleteItineraryPayload,
} from '../types';
import { addNotification } from './notificationsSlice';
import collectionFormItineraryDataMock from '../../../__mocks__/collectionFormItineraryDataMock';
import collectionFormItineraryAssetCategoryDataMock from '../../../__mocks__/collectionFormItineraryAssetCategoryDataMock';

/**
 * State slice definition for Collection Form Itinerary
 *
 * All state updates to the Collection Form Itinerary are managed here.
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 */

/**
 * Async thunk for GET /api/collection/itinerary/api/itinerary API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 * @memberof CollectionFormItineraryReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const fetchByCollectionId = createAsyncThunk(
  'collectionFormItinerary/fetchByCollectionId',
  async (args: IFetchCollectionFormItineraryByCollectionIdArgs, thunkAPI) => {
    /* await the result from the GET request */
    const res = await collectionFormService.getItinerary(args.collectionId);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to load Collection Form Itinerary from server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Async thunk for POST /api/collection/itinerary/api/itinerary API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 * @memberof CollectionFormItineraryReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const saveByCollectionId = createAsyncThunk(
  'collectionFormItinerary/saveByCollectionId',
  async (args: ISaveCollectionFormItineraryByCollectionIdArgs, thunkAPI) => {
    const res = await collectionFormService.setItinerary(args.data);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to save Collection Form Itinerary to server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    } else {
      thunkAPI.dispatch(
        addNotification({
          message: 'Successfully saved Collection Form Itinerary to server',
          variant: 'success',
        })
      );
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Async thunk for GET /api/collection/enumerations/api/asset_category API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 * @memberof CollectionFormItineraryReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const fetchTypes = createAsyncThunk(
  'collectionFormItinerary/fetchAssetCategories',
  async (args, thunkAPI) => {
    /* await the result from the GET request */
    const res = await collectionFormService.getItineraryAssetCategories();

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to load Itinerary Asset Category data from server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Initialises Collection Form Itinerary state to empty array
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 * @memberof CollectionFormItineraryReduxSlice
 *
 * @constant
 * @type {ICollectionFormItineraryState}
 */
const initialCollectionFormItineraryState: ICollectionFormItineraryState = {
  loading: false,
  data: collectionFormItineraryDataMock,
  dataShadow: collectionFormItineraryDataMock,
  saving: false,
  edited: false,
  loadingAssetCategories: false,
  assetCategories: collectionFormItineraryAssetCategoryDataMock,
};

/**
 * Create the redux slice for interacting with the Collection Form Itinerary state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 * @memberof CollectionFormItineraryReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {Slice<Object>}
 */
const collectionFormItinerarySlice = createSlice({
  name: 'collectionFormItinerary',
  initialState: initialCollectionFormItineraryState,
  reducers: {
    /* reducer used for user input changes to the Itinerary data */
    editItinerary: (
      state,
      action: PayloadAction<ICollectionFormItineraryEditItineraryPayload>
    ) => {
      /* find and update passed cost */
      state.data = state.data.map(
        (itinerary: ICollectionFormItineraryData, index: number) => {
          /* perform update for passed table row number */
          if (index === action.payload.rowIdx) {
            return {
              ...itinerary,
              /* update the value of the passed column */
              [action.payload.colKey as keyof ICollectionFormItineraryData]:
                action.payload.value,
            };
          }
          return itinerary;
        }
      );
      state.edited = true;
    },
    /* reducer used for when user adds Itinerary data */
    addItinerary: (state) => {
      /* add blank row to data */
      state.data = state.data.concat({
        asset_category: state.assetCategories[0],
        packing_details: '',
        quantity: 0,
      });
      state.edited = true;
    },
    /* reducer used for when user deletes Itinerary data */
    deleteItinerary: (
      state,
      action: PayloadAction<ICollectionFormItineraryDeleteItineraryPayload>
    ) => {
      // @ts-ignore
      state.data = state.data.toSpliced(action.payload.rowIdx, 1);
      state.edited = true;
    },
    /* reducer used for when user clears edits to Itinerary data */
    resetItinerary: (state) => {
      /* reset the data by simply copying the shadow to working copy */
      state.data = state.dataShadow;
      state.edited = false;
    },
  },
  extraReducers: (
    builder: ActionReducerMapBuilder<ICollectionFormItineraryState>
  ) => {
    /*
     * `extraReducers` used to handle actions that were generated
     * outside of the Collection Form Itinerary slice, such as the async thunks
     * for the API requests, defined at the top of this file.
     */
    builder
      /* Fetch Collection Form Itinerary extra reducers */
      .addCase(
        fetchByCollectionId.pending,
        (state: ICollectionFormItineraryState) => {
          state.loading = true;
          state.data = [];
          state.dataShadow = [];
          state.error = undefined;
          state.edited = false;
        }
      )
      .addCase(
        fetchByCollectionId.fulfilled,
        (
          state: ICollectionFormItineraryState,
          action: PayloadAction<ICollectionFormItineraryDataPayload>
        ) => {
          state.loading = false;
          state.data = action.payload;
          state.dataShadow = action.payload;
          state.error = undefined;
        }
      )
      .addCase(
        fetchByCollectionId.rejected,
        (state: ICollectionFormItineraryState) => {
          state.loading = false;
          state.data = collectionFormItineraryDataMock;
          state.dataShadow = collectionFormItineraryDataMock;
          state.error =
            'Failed to load Collection Form Itinerary data from server.';
        }
      )
      /* Save Collection Form Itinerary extra reducers */
      .addCase(
        saveByCollectionId.pending,
        (state: ICollectionFormItineraryState) => {
          state.saving = true;
        }
      )
      .addCase(
        saveByCollectionId.fulfilled,
        (state: ICollectionFormItineraryState) => {
          state.saving = false;
          state.edited = false;
        }
      )
      .addCase(
        saveByCollectionId.rejected,
        (state: ICollectionFormItineraryState) => {
          state.saving = false;
        }
      )
      /* Fetch Collection Form Itinerary Asset Category extra reducers */
      .addCase(fetchTypes.pending, (state: ICollectionFormItineraryState) => {
        state.loadingAssetCategories = true;
      })
      .addCase(
        fetchTypes.fulfilled,
        (
          state: ICollectionFormItineraryState,
          action: PayloadAction<ICollectionFormItineraryAssetCategoryDataPayload>
        ) => {
          state.loadingAssetCategories = false;
          state.assetCategories = action.payload;
        }
      )
      .addCase(fetchTypes.rejected, (state: ICollectionFormItineraryState) => {
        state.loadingAssetCategories = false;
        state.assetCategories = collectionFormItineraryAssetCategoryDataMock;
        state.error =
          'Failed to load Collection Form Itinerary Asset Category data from server.';
      });
  },
});

/* Collection Form Itinerary actions for editing and resetting itinerary data */
export const { editItinerary, addItinerary, deleteItinerary, resetItinerary } =
  collectionFormItinerarySlice.actions;

export default collectionFormItinerarySlice.reducer;
