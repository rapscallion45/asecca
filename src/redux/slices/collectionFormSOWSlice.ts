import {
  ActionReducerMapBuilder,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from '@reduxjs/toolkit';
import { ICollectionFormSOWDataPayload } from '@/lib/api/api-types';
import collectionFormService from '../../services/forms/collectionFormService';
import {
  ICollectionFormSOWState,
  IFetchCollectionFormSOWByCollectionIdArgs,
} from '../types';
import { addNotification } from './notificationsSlice';
import collectionFormSOWDataMock from '../../../__mocks__/CollectionForm/collectionFormSOWDataMock';

/**
 * State slice definition for Collection Form SOW
 *
 * All state updates to the Collection Form SOW are managed here.
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.21
 */

/**
 * Async thunk for GET /api/collection/sow/preview API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.21
 * @memberof CollectionFormSOWReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const fetchByCollectionId = createAsyncThunk(
  'collectionFormSOW/fetchByCollectionId',
  async (args: IFetchCollectionFormSOWByCollectionIdArgs, thunkAPI) => {
    /* await the result from the GET request */
    const res = await collectionFormService.getSOW(args.collectionId);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to load Collection Form SOW from server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Async thunk for GET /api/collection/sow/api/download API handling
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.21
 * @memberof CollectionFormSOWReduxSlice
 *
 * @see See [more info on Redux Async Thunks](https://redux-toolkit.js.org/api/createAsyncThunk)
 *
 * @function
 */
export const downloadByCollectionId = createAsyncThunk(
  'collectionFormSOW/downloadByCollectionId',
  async (args: IFetchCollectionFormSOWByCollectionIdArgs, thunkAPI) => {
    /* await the result from the GET request */
    const res = await collectionFormService.downloadSOW(args.collectionId);

    /* add a notification and reject if bad response from server */
    if (res.status !== 200) {
      thunkAPI.dispatch(
        addNotification({
          message: `Failed to download Collection Form SOW from server: ${res.statusText}`,
          variant: 'error',
        })
      );
      throw new Error(res.statusText);
    }

    /* no error, serialize the data and return */
    return res.json();
  }
);

/**
 * Initialises Collection Form SOW state to empty array
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.21
 * @memberof CollectionFormSOWReduxSlice
 *
 * @constant
 * @type {ICollectionFormSOWState}
 */
const initialCollectionFormSOWState: ICollectionFormSOWState = {
  loading: false,
  data: collectionFormSOWDataMock,
  downloading: false,
};

/**
 * Create the redux slice for interacting with the Collection Form SOW state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.21
 * @memberof CollectionFormSOWReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {Slice<Object>}
 */
const collectionFormSOWSlice = createSlice({
  name: 'collectionFormSOW',
  initialState: initialCollectionFormSOWState,
  reducers: {},
  extraReducers: (
    builder: ActionReducerMapBuilder<ICollectionFormSOWState>
  ) => {
    /*
     * `extraReducers` used to handle actions that were generated
     * outside of the Collection Form SOW slice, such as the async thunks
     * for the API requests, defined at the top of this file.
     */
    builder
      /* Fetch Collection Form SOW extra reducers */
      .addCase(
        fetchByCollectionId.pending,
        (state: ICollectionFormSOWState) => {
          state.loading = true;
          state.data = { pdf: '' };
          state.error = undefined;
        }
      )
      .addCase(
        fetchByCollectionId.fulfilled,
        (
          state: ICollectionFormSOWState,
          action: PayloadAction<ICollectionFormSOWDataPayload>
        ) => {
          state.loading = false;
          state.data = action.payload;
          state.error = undefined;
        }
      )
      .addCase(
        fetchByCollectionId.rejected,
        (state: ICollectionFormSOWState) => {
          state.loading = false;
          state.data = collectionFormSOWDataMock;
          state.error = 'Failed to load Collection Form SOW data from server.';
        }
      )
      /* Download Collection Form SOW extra reducers */
      .addCase(
        downloadByCollectionId.pending,
        (state: ICollectionFormSOWState) => {
          state.downloading = true;
        }
      )
      .addCase(
        downloadByCollectionId.fulfilled,
        (state: ICollectionFormSOWState) => {
          state.downloading = false;
        }
      )
      .addCase(
        downloadByCollectionId.rejected,
        (state: ICollectionFormSOWState) => {
          state.downloading = false;
          state.downloadingError = 'Failed to download SOW from server.';
        }
      );
  },
});

export default collectionFormSOWSlice.reducer;
