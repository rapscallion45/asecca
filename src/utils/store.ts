/* redux store configuration placeholder */
const useStore = () => {};
export default useStore;
