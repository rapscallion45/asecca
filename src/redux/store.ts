import { configureStore, Action } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import logger from 'redux-logger';
import userPermissionReducer from './slices/userPermissionSlice';
import costsConfigReducer from './slices/costsConfigSlice';
import notificationsReducer from './slices/notificationsSlice';
import themeReducer from './slices/themeSlice';
import { collectionsKanbanSlice } from './slices/kanbanSlice';

/**
 * Configuration of the redux store to manage the global application state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on Redux Store object](https://redux.js.org/api/store)
 *
 * @constant
 * @type {ToolkitStore<Object>}
 */
const store = configureStore({
  reducer: {
    userPermission: userPermissionReducer,
    costsConfig: costsConfigReducer,
    notifications: notificationsReducer,
    theme: themeReducer,
    collectionsKanban: collectionsKanbanSlice.reducer,
  },
  // @ts-ignore
  middleware:
    process.env.NODE_ENV === 'development'
      ? (getDefaultMiddleware) => getDefaultMiddleware().concat(logger)
      : (getDefaultMiddleware) => getDefaultMiddleware(),
});

/**
 * Shorthand application store dispatch type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppDispatch - application specific store dispatch type
 */
export type AppDispatch = typeof store.dispatch;

/**
 * Shorthand application state type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppState - application specific state object shape
 */
export type AppState = ReturnType<typeof store.getState>;

/**
 * Shorthand application Thunk Action type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppThunk - application specific Thunk action definition
 */
export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>;

export default store;
