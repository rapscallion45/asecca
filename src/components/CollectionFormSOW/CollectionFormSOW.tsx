import { FC, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppState, AppDispatch } from '@/redux/store';
import {
  Card,
  CardHeader,
  CardContent,
  Box,
  useMediaQuery,
  Typography,
} from '@mui/material';
import ErrorIcon from '@mui/icons-material/Error';
import WarningIcon from '@mui/icons-material/Warning';
import { LoadingButton } from '@mui/lab';
import { downloadByCollectionId } from '@/redux/slices/collectionFormSOWSlice';

/**
 * Collection Form SOW Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.21
 *
 * @typedef ICollectionFormSOWProps
 * @prop {string} collectionId - ID string of Collection for data API call
 */
interface ICollectionFormSOWProps {
  collectionId: string;
}

/**
 * Collection Form SOW
 *
 * Presents the Collection Form SOW PDF file to the user, populated with data
 * fetched from API: /api/collection/sow/preview
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.21
 *
 * @component
 * @param {ICollectionFormSOWProps} props - component props
 * @returns {FC} - collection form sow functional component
 */
const CollectionFormSOW: FC<ICollectionFormSOWProps> = (props) => {
  const { collectionId } = props;
  const isMobile = useMediaQuery('(min-width:600px)');

  /* shorthand helper for dispatching redux actions */
  const dispatch = useDispatch<AppDispatch>();

  /* get collection form SOW data held in redux state */
  const {
    data: SOWData,
    loading,
    error,
    downloading,
    valid,
  } = useSelector((state: AppState) => state.collectionFormSOW);

  /**
   * Handles the downloading of the SOW PDF
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.21
   *
   * @method
   */
  const handleDownload = useCallback((): void => {
    dispatch(downloadByCollectionId({ collectionId }));
  }, [collectionId, dispatch]);

  return (
    <Card>
      <CardHeader title="SOW" />
      <CardContent sx={{ pt: 0 }}>
        <Box
          sx={{
            marginTop: 2,
            display: 'flex',
            flexDirection: 'column',
            justifyContent: 'left',
          }}
        >
          <Box mb={2}>
            {valid.sow_validity === 'valid - generated' && (
              <iframe
                src={`data:application/pdf;base64,${SOWData.pdf}`}
                title={`SOW PDF Preview - Collection ${collectionId}`}
                width="100%"
                height="500px"
              />
            )}
            {valid.sow_validity === 'valid - not generated' && (
              <Box
                display="flex"
                flexDirection="column"
                py={7}
                alignItems="center"
              >
                <WarningIcon color="warning" fontSize="medium" />
                <Typography variant="body1" mt={2}>
                  No PDF generated for this SOW.
                </Typography>
              </Box>
            )}
            {valid.sow_validity === 'invalid' && (
              <Box
                display="flex"
                flexDirection="column"
                py={7}
                alignItems="center"
              >
                <ErrorIcon color="error" fontSize="medium" />
                <Typography variant="body1" mt={2}>
                  SOW Invalid.
                </Typography>
              </Box>
            )}
          </Box>
          <LoadingButton
            color="secondary"
            variant="contained"
            onClick={handleDownload}
            disabled={
              downloading ||
              loading ||
              Boolean(error) ||
              valid.sow_validity !== 'valid - generated'
            }
            loading={downloading}
            fullWidth={isMobile}
            sx={{ width: isMobile ? '200px' : '100%' }}
          >
            Download PDF
          </LoadingButton>
        </Box>
      </CardContent>
    </Card>
  );
};

export default CollectionFormSOW;
