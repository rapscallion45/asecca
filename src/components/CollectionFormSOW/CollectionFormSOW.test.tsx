import React from 'react';
import { Provider } from 'react-redux';
import {
  fireEvent,
  render,
  screen,
  waitFor,
  act,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import { fetchByCollectionId as fetchCollectionFormSOWByCollectionId } from '@/redux/slices/collectionFormSOWSlice';
import store from '@/redux/store';
import CollectionFormSOW from './CollectionFormSOW';

/* default test query ID */
const query: string = '66135000001760012';

/**
 * Collection Form SOW Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Collection Form SOW', () => {
  /** mock iframe component */
  beforeEach(() => {
    jest.resetModules();
    global.document = {
      ...global.document,
      querySelector: () => ({
        addEventListener: jest.fn(),
      }),
    };
  });

  it('Renders correctly', async () => {
    /** Arrange */
    store.dispatch(
      fetchCollectionFormSOWByCollectionId({
        collectionId: query,
      })
    );

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <CollectionFormSOW collectionId={query} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Form Callbacks', () => {
    beforeAll(() => jest.spyOn(window, 'fetch'));

    it('Should call download API with with collection ID when button clicked', async () => {
      /** Arrange */

      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormSOW collectionId={query} />
        </Provider>
      );

      /** Asset - check the download button is rendered */
      await waitFor(() => {
        expect(screen.getByText('Download PDF')).toBeInTheDocument();
      });

      /** Act - click Download button */
      act(() => {
        fireEvent(
          screen.getByText('Download PDF'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });

      /** Assert - download API to have been called with passed collection ID */
      await waitFor(() => {
        expect(window.fetch).toHaveBeenCalledTimes(1);
      });
      expect(window.fetch).toHaveBeenCalledWith(
        `/api/collection/sow/api/download?collection=${query}`,
        expect.objectContaining({
          method: 'GET',
          headers: { 'Content-Type': 'application/json' },
        })
      );
    });

    // it('Should not render download button if Valid API has returned invalid', async () => {
    //   /** Arrange */

    //   /** Act - render the test components */
    //   render(
    //     <Provider store={store}>
    //       <CollectionFormSOW collectionId={query} />
    //     </Provider>
    //   );

    //   /** Asset - check the download button is rendered */
    //   await waitFor(() => {
    //     expect(screen.getByText('Download PDF')).toBeInTheDocument();
    //   });

    //   /** Act - click Download button */
    //   act(() => {
    //     fireEvent(
    //       screen.getByText('Download PDF'),
    //       new MouseEvent('click', {
    //         bubbles: true,
    //         cancelable: true,
    //       })
    //     );
    //   });

    //   /** Assert - download API should not have been called and error msg shown */
    //   await waitFor(() => {
    //     expect(window.fetch).toHaveBeenCalledTimes(0);
    //   });
    //   expect(screen.getAllByText('SOW Invalid.'));
    // });
  });
});
