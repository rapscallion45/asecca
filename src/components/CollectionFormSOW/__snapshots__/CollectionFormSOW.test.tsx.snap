// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Form SOW Renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
      >
        SOW
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root css-1hizqsf-MuiCardContent-root"
  >
    <div
      className="MuiBox-root css-37r54y"
    >
      <div
        className="MuiBox-root css-1qm1lh"
      >
        <iframe
          height="500px"
          src="data:application/pdf;base64,"
          title="SOW PDF Preview - Collection 66135000001760012"
          width="100%"
        />
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-8fpfx2-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        disabled={true}
        id=":r0:"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Download PDF
      </button>
    </div>
  </div>
</div>
`;
